base/operators.jl

```diff
@@ -856,6 +859,8 @@ julia> ∘(fs...)(3)
 3.0
 ```
 """
+∘() = identity
```

Comment:
I wonder if this is the best way.  For example, `∘(f, g, ∘())` may not work out-of-the-box if `f` and `g` are generic morphisms like lenses or transducers.  It's the same reason why we don't have `+()` or `*()`.  OTOH, it's nice that we have a canonical identity for `∘`.  I guess I'd be less worried if

```julia
∘(f, ::typeof(identity)) = f
∘(::typeof(identity), f) = f
∘(::typeof(identity), ::typeof(identity)) = identity
```

are implemented as well.

Reply:
That is a very good observation. My favored solution would be dropping `∘()` and adding your remark as a comment I think.

Reply:
Or perhaps `∘()` = error(your remark)`.

Reply:
That's based on the desire to have `∘` for other "morphisms" than functions, which has been requested before, e.g. in the discussion where `compose` was requested as an ASCII name for `∘`.

Reply:
Adding only `∘(f)` sounds good.  Adding things is much easier than removing things.

Reply:
@StefanKarpinski Do you mean that it's not recommended to use `∘` for other morphisms anymore now that #33573 was closed?  That's why we should keep `∘() = identity`?

Reply:
I'm just noting that not everyone agrees that `∘` should be considered a general operator for composition of morphisms. And, as you've pointed out, even if it is, we could use `identity` as the identity morphism. But I do suppose the most conservative thing to do is not define the zero-argument composition until it becomes clear that it's needed.

Reply:
Thanks for the explanation. I agree that not adding `∘()` is a conservative choice and has the extra benefit of avoiding to touch the controversial point.

Reply:
I would also be fine with just adding the definitions you proposed:
```
∘(f, ::typeof(identity)) = f
∘(::typeof(identity), f) = f
∘(::typeof(identity), ::typeof(identity)) = identity
```

Reply:
I agree. I'm also fine with both choices.

I actually like adding that three-line definition with `::typeof(identity)`.  It makes writing morphisms slightly easier.  But this can happen independently of `∘()`. If we need some excuse for adding that three-line definition, and if `∘()` is enough for that, then I'd vote for adding `∘()`. Yeah, that's a bit :smiling_imp:

Reply:
While I can live with both choices, but here are some arguments why less is more:

I take lens composition as an example, but this applies to transducers and probably most kinds of morphisms.

* When I write `lens = ∘(lenses...)` I expect that I can use the lens api on the result. Like `get(obj, lens)`. This is impossible to achieve for the empty case. Now defining `∘() = identity` (along with @tkf clever extra methods say) means that some part of the lens will work e.g `∘(lens, ∘())` and others will not. 
I would favor if everything fails at composition time with a sane error and not having some stuff work and other stuff throw a latend error about using lens api on `identity`.
* Morphisms often ship their own identity (since you want an identity of type `::Lens`). In order to make things work, one must also define e.g.
```julia
∘(::IdenityLens, typeof(∘()))
∘(typeof(∘()), ::IdenityLens)
```
Not the end of the world, but having two identities is an extra stumbling block package authors need to keep in mind.

Reply:
These are very good points. I'm now against `∘()`.

Reply:
Ok, so let's go with the less option then: remove the `∘()` and this should be good to go.